pi_top_usb_setup/network.py

```diff
@@ -0,0 +1,454 @@
+#################################
+@dataclass
+class WpaPersonal(NetworkBase):
+    password: str
+
+    def to_nmcli(self) -> str:
+        return f"mode infa \
+802-11-wireless-security.key-mgmt wpa-psk \
+802-11-wireless-security.psk {self.password}"
+
+    def to_wpasupplicant_conf(self) -> List:
+        return [
+            "key-mgmt=WPA-PSK",
+            f'psk="{self.password}"',
+        ]
+
+
+@dataclass
+class LEAP(NetworkBase):
+    username: str
+    password: Optional[str] = None
+
+    def to_nmcli(self) -> str:
+        response = f"802-11-wireless-security.auth-alg leap \
+802-11-wireless-security.key-mgmt ieee8021x \
+802-11-wireless-security.leap-username {self.username}"
+        if self.password:
+            response += f"802-11-wireless-security.leap-password {self.password}"
+        return response
+
+    def to_wpasupplicant_conf(self) -> List:
+        response = [
+            "auth_alg=LEAP",
+            "key_mgmt=IEEE8021X",
+            f'identity="{self.username}"',
+        ]
+        if self.password:
+            response.append(f'password="{self.password}"')
+        return response
+
+
+@dataclass
+class OWE:
+    def to_nmcli(self) -> str:
+        return "802-11-wireless-security.key-mgmt OWE"
+
+    def to_wpasupplicant_conf(self) -> List:
+        return ["key_mgmt=OWE"]
+
+
+@dataclass()
+class WpaEnterprise(NetworkBase):
+    authentication: (
+        PWDAuthentication | TLSAuthentication | TTLSAuthentication | PEAPAuthentication
+    )
+
+    @classmethod
+    def from_kwargs(cls, **kwargs):
+        print(f"--------> WpaEnterprise: {kwargs}")
+        authentication_lookup = {
+            WpaEnterpriseAuthentication.PWD: PWDAuthentication,
+            WpaEnterpriseAuthentication.TLS: TLSAuthentication,
+            WpaEnterpriseAuthentication.TTLS: TTLSAuthentication,
+            WpaEnterpriseAuthentication.PEAP: PEAPAuthentication,
+        }
+        try:
+            authentication_type = kwargs["authentication"]
+            authentication_enum = WpaEnterpriseAuthentication[authentication_type]
+            authentication_cls = authentication_lookup[authentication_enum]
+        except KeyError:
+            raise ValueError(f"Invalid authentication type '{authentication_type}'")
+
+        return WpaEnterprise(authentication=authentication_cls.from_kwargs(**kwargs))
+
+    def to_nmcli(self) -> str:
+        return self.authentication.to_nmcli()
+
+    def to_wpasupplicant_conf(self) -> List:
+        return self.authentication.to_wpasupplicant_conf()
+
+
+#################################
+# Main class
+#################################
+
+
+@dataclass
+class Network:
+    ssid: str
+    authentication: WpaPersonal | LEAP | OWE | WpaEnterprise
+    hidden: bool = False
+
+    def connect(self):
+        if not (
+            isinstance(self.authentication, WpaPersonal)
+            or isinstance(self.authentication, OWE)
+        ):
+            raise NotImplementedError(
```

Comment:
IMO the easiest way to get this tested is to have it implemented here and just go out and see when it breaks - obviously ideally we can test internally not have customers do it.

maybe consider adding very good logging so when we have a test log we can work it out